content/blog/2022/2022-03-31-be-informed-with-lombok.md

```diff
@@ -0,0 +1,474 @@
+
+### @SneakyThrows can be evil
+
+Let's first consider this example:
+````java
+   public interface DataProcessor {
+       void dataProcess();
+   }   
+````
+Without @SneakyThrows an implementation of DataProcessor would be like this:
+````java
+   public class FileDataProcessor implements DataProcessor {
+       @Override
+       public void dataProcess() {
+           try {
+               processFile();
+           } catch (IOException e) {
+               e.printStackTrace();
+           }
+       }
+   
+       private void processFile() throws IOException {
+           File file = new File("sample.txt");
+           throw new IOException(); // forcibly throw
+       }
+   }
+````
+With @SneakyThrows the code gets simplified
+````java
+   public class FileDataProcessor implements DataProcessor {
+       @Override
+       @SneakyThrows
+       public void dataProcess() {
+          processFile();
+       }
+   
+       private void processFile() throws IOException {
+           File file = new File("sample.txt");
+           throw new IOException();
+       }
+   }
+````
+As we can see, this **avoids the hassle of catching or throwing checked exceptions**. In other words, it treats a checked exception like an unchecked one.
+This can be useful, especially when writing lambda functions making the code concise and clean. However, since the annotation swallows the checked exception we cannot catch them explicitly. 
+Instead, we must handle the exceptions via a global exception handler.
+Therefore, **use it only when you don't intend to process the code selectively depending on the kind of Exception it throws**.
+**Ensure it is used cautiously and not as an alternative to bypass checked exceptions.**
+
+## Use Lombok with caution
+
+The power of Lombok cannot be underestimated or ignored. However, I would like to summarise the key points
+that will help you use Lombok in a better way.
+1. **Avoid using Lombok with entities**. It will be much easier generating the code yourself than debugging issues later.
+2. When designing POJO's **use only the Lombok annotations you require**. (Use shorthand annotations wisely).
+I would recommend using the Delombok feature to understand the code generated better.
+3. **Do not add too many dependencies in the POJO's**. Keep the classes relevant to the responsibility they are designed for. 
+It is easy to lose track of dependent objects with Lombok.
+4. Since @Builder gives a lot of flexibility in object creation it **can cause objects to be in an invalid state**. 
+Therefore, make sure all the required attributes are assigned values during object creation.
+5. **DO NOT write code that could have a huge dependency on the background code Lombok generates**.
+6. When using test coverage tools like Jacoco, Lombok can cause problems since **Jacoco cannot distinguish between Lombok generated code and normal source code**. 
```

Comment:
Good point! You haven't mentioned JaCoCo above! This might be worth to mention in the section about validations (maybe rephrase the title of that section to something like "Using Lombok with Static Code Analyzers"

Reply:
@thombergs  I see that the website already has an aricle about Jacoco that mentions the steps to exclude the code generated by Lombok. Do I just add a link to this in the main section or should I add some details on why this is reqd and then add a link to this section?
https://reflectoring.io/jacoco/#excluding-code-generated-by-lombok

Reply:
Just mention it and add a link